### PR TITLE
[NPU] update save-load API usage

### PR DESCRIPTION
**Pull request by plusbang**
## Description

Update save-load API usage.

- save: For all cases with `optimize_model=True`, specify `save_directory` is required during the first time to load model, ckpt is saved to `save_directory` after converting.
- load: still use `load_low_bit` API to load, and add python cpp backend support

### 2. User API changes
```python
# first time load, and ckpt is saved to `save_directory` after converting
model = AutoModelForCausalLM.from_pretrained(
    model_path,
    torch_dtype=torch.float16,
    trust_remote_code=True,
    attn_implementation="eager",
    optimize_model=True,
    max_context_len=1024,
    max_prompt_len=960,
    mixed_precision=True,
    quantization_group_size=0,
    save_directory=save_directory  # required and has related check
)

# load converted model
model = AutoModelForCausalLM.load_low_bit(
    save_directory,
    attn_implementation="eager",
    torch_dtype=torch.float16,
    optimize_model=True,
    max_context_len=1024,
    max_prompt_len=960,
    trust_remote_code=True,
)
```
example and benchmark scripts have updated.

### 4. How to test?
- [x] Application test

### CODE REVIEW

File: python/llm/src/ipex_llm/transformers/npu_model.py [hunk 1]
```diff
@@ -45,6 +45,9 @@ def ignore_argument(kwargs: dict, key: "str"):
 
 
 def save_low_bit(self, model_dir: str, *args, **kwargs):
+    if hasattr(self, "save_directory"):
+        logger.info(f"Model has already saved to {self.save_directory}.")
+        return 1
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/npu_model.py:50]: What if "save_directory" is different from "model_dir"?

File: python/llm/src/ipex_llm/transformers/npu_model.py [hunk 2]
```diff
@@ -267,6 +270,9 @@ def optimize_npu_model(cls, *args, **kwargs):
 
         if not pipeline:
             if model.config.model_type in ["qwen2", "llama", "minicpm"]:
+                invalidInputError(save_directory is not None,
+                                  "Please provide the path to save converted model "
+                                  "through `save_directory`.")
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/npu_model.py:275]: Can we use temp dir in this case?

File: python/llm/src/ipex_llm/transformers/npu_model.py (hunk 1 above)
Comment (plusbang) [reply python/llm/src/ipex_llm/transformers/npu_model.py]: > What if "save_directory" is different from "model_dir"?

Have updated the logic of save.

For all cases with `optimize_model=True`, specify `save_directory` is required during the first time to load model, ckpt is saved to `save_directory` after converting.

File: python/llm/src/ipex_llm/transformers/npu_model.py [hunk 3]
```diff
@@ -45,6 +45,8 @@ def ignore_argument(kwargs: dict, key: "str"):
 
 
 def save_low_bit(self, model_dir: str, *args, **kwargs):
+    if hasattr(self, "save_directory"):
```
Comment (jason-dai) [inline python/llm/src/ipex_llm/transformers/npu_model.py:48]: Report a warning that the model is already saved at "save_directory"

**Review by rnwang04 — APPROVED** [review]
LGTM

**Comment by plusbang** [comment]
Merge it first and will add tokenizer save-load processing in examples later.

**Pull request merged**